File: go/domemo.go
//
// domemo.go
//
// Copyright (c) 2018 Hiroaki Wada
//
// This software is released under the MIT License.
// http://opensource.org/licenses/mit-license.php
//
package main

import(
	"fmt"
	"math/rand"
	"sort"
)

type Player struct {
	name string
	cards []int
}

func (p Player)check() bool {
	return len(p.cards) == 0
}

func (p Player) show() {
	fmt.Println(p.name + " -> ", p.cards)
}

func match(p *Player, guessCard int) bool {
	ret := false
	cardLen := len(p.cards)
	for idx := 0; idx < cardLen; idx++ {
		if p.cards[idx] == guessCard {
			fmt.Println("jackpot!")
			ret = true
			if idx == cardLen{
				p.cards = p.cards[0:idx-1]
			}else if idx == 0 {
				p.cards = p.cards[idx+1:cardLen]
			}else{
				p.cards = append(p.cards[0:idx], p.cards[idx+1:cardLen]...)
			}
			break;
		}
	}

	if !ret {
		fmt.Println("miss!")
	}

	return ret
}

func main(){
	cards := [28]int{1,2,2,3,3,3,4,4,4,4,5,5,5,5,5,6,6,6,6,6,6,7,7,7,7,7,7,7}
	cardLen := len(cards)
	for idx := 0; idx < cardLen; idx++ {
		target := rand.Intn(cardLen)
		swp := cards[target]
		cards[target] = cards[idx]
		cards[idx] = swp
	}

	openCards := cards[cardLen-4:cardLen]
	players := []Player{}

	for idx := 0; idx < 4; idx++ {
		name := "Player" + fmt.Sprintf("%d", idx+1)
		playerCards := cards[5*idx:5*(idx+1)]
		sort.Sort(sort.IntSlice(playerCards))
		players = append(players, Player{name, playerCards})
	}

	round := 1
GAME_LOOP:
	for round < 30 {
		fmt.Println("*round", round)
		fmt.Println("opencards -> ", openCards)
		for idx := 0; idx < 4; idx++ {
			players[idx].show()
		}

		for idx := 0; idx < 4; idx++ {
			guessCard := rand.Intn(7) + 1
			fmt.Println(players[idx].name, " is Called -> ", guessCard)
			if match(&players[idx], guessCard){
				openCards = append(openCards, guessCard)
				sort.Sort(sort.IntSlice(openCards))
			}
			if players[idx].check(){
				fmt.Println(players[idx].name, " is Win!")
				break GAME_LOOP
			}
		}
		round++
	}
}

File: kotlin/domemo.kt
/*
  domemo.kt

  Copyright (c) 2018 Hiroaki Wada

  This software is released under the MIT License.
  http://opensource.org/licenses/mit-license.php
*/
import java.util.Random

class Player{
    val name: String
    val cards: MutableList<Int>

    constructor(name: String, cards: MutableList<Int>){
        this.name = name
        this.cards = cards
    }

    fun show(){
        println(this.name + " -> " + this.cards)
    }

    fun match(guessCard: Int): Boolean {
        val ret = cards.remove(guessCard)
        if(ret){
            println("jackpot!")
        }else{
            println("miss!")
        }
        return ret
    }

    fun check(): Boolean{
        return this.cards.count() == 0 
    }
}

fun main(args: Array<String>) {
    val cards = mutableListOf(1,2,2,3,3,3,4,4,4,4,5,5,5,5,5,6,6,6,6,6,6,7,7,7,7,7,7,7)
    cards.shuffle()

    val sliceCards = cards.slice(cards.count()-4 .. cards.count()-1).sorted()
    var openCards: MutableList<Int> = sliceCards.toMutableList()

    val players: MutableList<Player> = mutableListOf()
    for(i in 0 until 4){
        var playerCards = cards.slice(i*5 .. (i+1)*5-1).sorted()
        players.add(Player("Player"+(i+1), playerCards.toMutableList()))
    }

    var round = 1
    while(round < 30){
        println("*round" + round)
        println(openCards)
        for(player in players){
            player.show()
        }

        for(player in players){
            val random: Random = Random()
            val guessCard = random.nextInt(7)+1
            println(player.name + " is Called -> " + guessCard)

            if(player.match(guessCard)){
                openCards.add(guessCard)
                val sortedOpenCards = openCards.sorted()
                openCards = sortedOpenCards.toMutableList()

                if(player.check()){
                    println(player.name + " is Win!")
                    return
                }
            }
        }
        round++
    }
}